pkg/ingress/gateway.go

```diff
@@ -134,58 +77,80 @@ func (c *client) handleCertificateChange(currentCertSpec *configv1alpha2.Ingress
 				continue
 			}
 			if newCertSpec == nil && currentCertSpec != nil {
+				stopWatching()
 				// Implies the certificate reference was removed, delete the corresponding secret and certificate
 				if err := c.removeGatewayCertAndSecret(*currentCertSpec); err != nil {
 					log.Error().Err(err).Msg("Error removing stale gateway certificate/secret")
 				}
-			} else if newCertSpec != nil {
-				// New cert spec is not nil and is not the same as the current cert spec, update required
-				err := c.createAndStoreGatewayCert(*newCertSpec)
-				if err != nil {
-					log.Error().Err(err).Msgf("Error updating ingress gateway cert and secret")
-				}
+			} else if shouldRelisten(currentCertSpec, newCertSpec) {
+				stopWatching()
+				stopWatching = c.watchCertificateRotations(newCertSpec.SubjectAltNames[0], newCertSpec.Secret)
 			}
 			currentCertSpec = newCertSpec
 
-		// A certificate was rotated
-		case msg, ok := <-certRotateChan:
-			if !ok {
-				log.Warn().Msg("Notification channel closed for certificate rotation")
-				continue
-			}
+		case <-stop:
+			stopWatching()
+			return
+		}
+	}
+}
 
-			event, ok := msg.(events.PubSubMessage)
-			if !ok {
-				log.Error().Msgf("Received unexpected message %T on channel, expected PubSubMessage", event)
-				continue
-			}
-			cert, ok := event.NewObj.(*certificate.Certificate)
-			if !ok {
-				log.Error().Msgf("Received unexpected message %T on cert rotation channel, expected Certificate", cert)
-				continue
-			}
+func shouldRelisten(oldSpec *configv1alpha2.IngressGatewayCertSpec, newSpec *configv1alpha2.IngressGatewayCertSpec) bool {
+	if newSpec == nil || len(newSpec.SubjectAltNames) == 0 {
+		return false
+	}
 
-			// This should never happen, but guarding against a panic due to the usage of a pointer
-			if currentCertSpec == nil {
-				log.Error().Msgf("Current ingress gateway cert spec is nil, but a certificate for it was rotated - unexpected")
-				continue
-			}
+	if oldSpec == nil || len(oldSpec.SubjectAltNames) == 0 {
+		return true
+	}
 
-			cnInCertSpec := currentCertSpec.SubjectAltNames[0] // Only single SAN is supported in certs
+	return oldSpec.SubjectAltNames[0] != newSpec.SubjectAltNames[0]
+}
 
-			// Only update the secret if the cert rotated matches the cert spec
-			if cert.GetCommonName() != certificate.CommonName(cnInCertSpec) {
-				continue
-			}
+// watchCertificateRotations creates and stores a certificate with the given common name into the given secret ref.
+// It then watches for all cert rotations and updates the secret on changes.
+func (c *client) watchCertificateRotations(cn string, secret corev1.SecretReference) func() {
```

Comment:
Why do we need `watchCertificateRotations`? Doesn't the existing pub/sub infrastructure do the same? 
Whoever needs to watch for cert rotation could subscribe to the `cert-rotated` topic, no?